blockchain/state_migration.go

```diff
@@ -528,3 +531,141 @@ func (bc *BlockChain) StopWarmUp() error {
+
+	startNode := block.Root()
+	// If the contractAddr is given, start collecting stats from the root of storage trie
+	if !common.EmptyAddress(contractAddr) {
+		var err error
+		startNode, err = bc.GetContractStorageRoot(block, db, contractAddr)
+		if err != nil {
+			logger.Error("Failed to get the contract storage root",
+				"contractAddr", contractAddr.String(), "rootHash", block.Root().String(),
+				"err", err)
+			return err
+		}
+	}
+
+	children, err := db.TrieDB().NodeChildren(startNode)
+	if err != nil {
+		logger.Error("Failed to retrieve the children of start node", "err", err)
+		return err
+	}
+
+	logger.Info("Started collecting trie statistics",
+		"blockNum", block.NumberU64(), "root", block.Root().String(), "len(children)", len(children))
+	go collectTrieStats(db, startNode)
+
+	return nil
+}
+
+// errFinishedCollecting is used to indicate that a collectChildrenStats call is finished.
+var errFinishedCollecting = errors.New("finished collecting stats")
+
+// collectChildrenStats wraps CollectChildrenStats, in order to send finish signal to resultCh.
+func collectChildrenStats(db state.Database, child common.Hash, resultCh chan<- statedb.NodeInfo) {
+	db.TrieDB().CollectChildrenStats(child, 2, resultCh)
+	resultCh <- statedb.NodeInfo{Err: errFinishedCollecting}
+}
+
+// collectTrieStats is the main function of collecting trie statistics.
+// It spawns goroutines for the upper-most children and
+func collectTrieStats(db state.Database, startNode common.Hash) {
+	children, err := db.TrieDB().NodeChildren(startNode)
+	if err != nil {
+		logger.Error("Failed to retrieve the children of start node", "err", err)
+	}
+
+	// collecting statistics by running individual goroutines for each child
+	resultCh := make(chan statedb.NodeInfo, 10000)
+	for _, child := range children {
+		go collectChildrenStats(db, child, resultCh)
+	}
+
+	numGoRoutines := len(children)
+	ticker := time.NewTicker(1 * time.Minute)
+
+	numNodes, numLeafNodes, maxDepth := 0, 0, 0
+	depthCounter := make(map[int]int)
+	begin := time.Now()
+	for {
+		select {
+		case result := <-resultCh:
+			if result.Err == errFinishedCollecting {
+				numGoRoutines--
+				continue
+			}
+			numNodes++
+			// if a leaf node, collect the depth data
+			if result.Depth != 0 {
+				numLeafNodes++
+				depthCounter[result.Depth]++
+				if result.Depth > maxDepth {
+					maxDepth = result.Depth
+				}
+			}
+		case <-ticker.C:
+			// leave a periodic log
+			logger.Info("Collecting trie statistics is in progress...", "elapsed", time.Since(begin),
+				"numGoRoutines", numGoRoutines, "numNodes", numNodes, "numLeafNodes", numLeafNodes, "maxDepth", maxDepth)
+			printDepthStats(depthCounter)
+		default:
+			if numGoRoutines != 0 {
+				continue
+			}
+			logger.Info("Finished collecting trie statistics", "elapsed", time.Since(begin),
+				"numNodes", numNodes, "numLeafNodes", numLeafNodes, "maxDepth", maxDepth)
+			printDepthStats(depthCounter)
+			return
```

Comment:
When there is no data, these lines will run infinitvely. How about putting these in line 609?

Reply:
That would be better! I will update the PR.